src/sentry/sentry_metrics/indexer/cache.py

```diff
@@ -136,14 +196,31 @@ def bulk_record(
 
         db_record_keys = cache_key_results.get_unmapped_keys(cache_keys)
 
+        overall_cache_results = cache_key_results.merge(new_cache_key_results)
+
         if db_record_keys.size == 0:
-            return cache_key_results
+            return overall_cache_results
 
         db_record_key_results = self.indexer.bulk_record(use_case_id, db_record_keys.mapping)
+
+        # We will temporarily change it to write new cache keys for all results
+
+        # DB key results
         self.cache.set_many(
             db_record_key_results.get_mapped_key_strings_to_ints(), use_case_id.value
         )
-        return cache_key_results.merge(db_record_key_results)
+
+        self.cache.set_many_new(
+            db_record_key_results.get_mapped_key_strings_to_ints(), use_case_id.value
+        )
+
+        # cache key results (these are on the old cache keys,
+        # and we need to record the new cache keys)
+        self.cache.set_many_new(
+            cache_key_results.get_mapped_key_strings_to_ints(), use_case_id.value
+        )
+
+        return overall_cache_results.merge(db_record_key_results)
```

Comment:
It looks like when the a bulk_record call comes in, we have:
- `cache_results`: The cache hits (from the original namespace)
- `db_record_key_results`: Results from querying postgres
- [NEW] `new_cache_results`: The cache hits (from the new namespace)
- [NEW] `new_cache_miss_keys `: The misses from the cache (from the new namespace)
- [NEW] `overall_cache_results`: The misses from the cache (from **all** namespaces)

Is it necessary to calculate  `overall_cache_results`? What is the downside to just doing:

```python3

self.cache.set_many_new(cache_results, use_case_id.value)

if db_record_keys.size == 0:
            return cache_key_results

self.cache.set_many_new(db_record_key_results. get_mapped_key_strings_to_ints())

return cache_key_results.merge(db_record_key_results)
```

Reply:
So, in order for our new cache keys (that we create via `set_many_new`) to not expire after 2 hours, we need to be reading them/retrieving them from the cache (a cache hit). If we simply "set" these keys as a regular operation while never actually hitting them, they will expire naturally, and we don't want that. 

So that's why those approaches wouldn't work. Does that make sense?

Reply:
`overall_cache_results` here tells you all the cache hits you had, not misses. And this function returns all the cache hits results + the db record results. So the reason why we have `overall_cache_results` is so that this function can return the old cache key hit results + the new cache key hit results + the db record key results.

Reply:
Ok, so I just learned that the eviction policy is simply remove cache keys after 2 hours. So that can help make this approach simpler. I'll update it in a bit